File: MangaViewer/app/src/main/java/com/king/mangaviewer/adapter/MangaChapterItemAdapter.java
package com.king.mangaviewer.adapter;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.king.mangaviewer.R;
import com.king.mangaviewer.activity.MangaPageActivity;
import com.king.mangaviewer.component.MyImageView;
import com.king.mangaviewer.util.AsyncImageLoader;
import com.king.mangaviewer.model.MangaChapterItem;
import com.king.mangaviewer.viewmodel.MangaViewModel;

import java.util.List;

public class MangaChapterItemAdapter extends RecyclerView.Adapter<MangaChapterItemAdapter.RecyclerViewHolders> {

    int MAX_TITLE_LENGTH = 20;
    protected Context context;
    protected LayoutInflater mInflater = null;
    protected MangaViewModel viewModel;
    protected List<MangaChapterItem> chapter;

    public MangaChapterItemAdapter(Context context, MangaViewModel viewModel,
                                   List<MangaChapterItem> chapter) {
        super();
        this.mInflater = LayoutInflater.from(context);
        this.viewModel = viewModel;
        this.context = context;
        this.chapter = chapter;

    }

    @Override
    public int getItemCount() {
        // TODO Auto-generated method stub
        return chapter.size();
    }

    @Override
    public RecyclerViewHolders onCreateViewHolder(ViewGroup parent, int viewType) {
        View layoutView = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_manga_chapter_item, parent, false);
        RecyclerViewHolders rcv = new RecyclerViewHolders(layoutView);
        return rcv;
    }

    @Override
    public void onBindViewHolder(RecyclerViewHolders holder, int position) {
        String chapterTitle = chapter.get(position).getTitle();
        if (chapter.get(position).getMenu().getTitle().length() > MAX_TITLE_LENGTH) {
            chapterTitle = chapterTitle.replace(chapter.get(position).getMenu().getTitle(), context.getString(R.string.prefix_chapter_title));
            chapter.get(position).setTitle(chapterTitle);
        }
        holder.textView.setText(chapterTitle);

    }


    public class RecyclerViewHolders extends RecyclerView.ViewHolder implements View.OnClickListener {

        public ImageView imageView;
        public TextView textView;

        public RecyclerViewHolders(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);

            imageView = (MyImageView) itemView.findViewById(R.id.imageView);
            textView = (TextView) itemView.findViewById(R.id.textView);
        }

        @Override
        public void onClick(View view) {
            int position = getPosition();
            // TODO Auto-generated method stub
            viewModel.setSelectedMangaChapterItem(chapter.get(position));
            context.startActivity(new Intent(context, MangaPageActivity.class));
            ((Activity) context).overridePendingTransition(R.anim.in_rightleft, R.anim.out_rightleft);

        }
    }
}

File: MangaViewer/app/src/main/java/com/king/mangaviewer/adapter/MangaChapterItemAdapter.kt
package com.king.mangaviewer.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.INVISIBLE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import com.king.mangaviewer.R
import com.king.mangaviewer.adapter.MangaChapterItemAdapter.RecyclerViewHolders
import com.king.mangaviewer.domain.data.local.DownloadState
import com.king.mangaviewer.domain.data.local.DownloadState.DOWNLOADING
import com.king.mangaviewer.domain.data.local.DownloadState.ERROR
import com.king.mangaviewer.domain.data.local.DownloadState.FINISHED
import com.king.mangaviewer.domain.data.local.DownloadState.NONE
import com.king.mangaviewer.domain.data.local.DownloadState.PENDING
import com.king.mangaviewer.model.MangaChapterItem

class MangaChapterItemAdapter(private val context: Context,
    private val onItemClickListener: OnItemClickListener,
    private val onSelectedChangeListener: OnSelectedChangeListener? = null) :
    BaseRecyclerViewAdapter<MangaChapterItem, RecyclerViewHolders>(diffCallBack) {

    private val MAX_TITLE_LENGTH = 20
    private var stateMap: Map<String, MangaChapterStateItem> = emptyMap()
    private var selectableMode = false
    private val selectedMap: MutableMap<Int, Boolean> = hashMapOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolders {
        val layoutView = LayoutInflater.from(parent.context).inflate(
            R.layout.list_manga_chapter_item, parent, false)
        return RecyclerViewHolders(layoutView)
    }

    override fun onBindViewHolder(holder: RecyclerViewHolders, position: Int) {
        val item = getItem(position)
        item?.apply {
            var chapterTitle = title
            if (menu.title.length > MAX_TITLE_LENGTH) {
                chapterTitle = chapterTitle.replace(menu.title,
                    context.getString(R.string.prefix_chapter_title))
                title = chapterTitle
            }
            holder.textView.text = chapterTitle
            holder.itemView.setOnClickListener {
                if (selectableMode) {
                    holder.cbDownload.toggle()
                    toggleSelected(position)
                } else {
                    onItemClickListener.onClick(this)
                }
            }
            holder.cbDownload.setOnClickListener {
                toggleSelected(position)
            }
            holder.ivState.setOnClickListener {
                //TODO resume or cancel
            }
            holder.setRead(stateMap[item.hash]?.isRead ?: false)
            holder.setDownloadState(stateMap[item.hash]?.downloaded ?: NONE)
            holder.setSelectable(selectableMode, selectedMap[position] ?: false)
        }
    }

    fun submitStateMap(map: Map<String, MangaChapterStateItem>) {
        stateMap = map
        notifyDataSetChanged()
    }

    fun toggleSelectableMode() {
        selectableMode = selectableMode.not()
        if (!selectableMode) {
            selectedMap.clear()
        }
        notifyDataSetChanged()
        notifySelectedChange()
    }

    private fun toggleSelected(position: Int) {
        val value = selectedMap[position] ?: false

        if (value) {
            selectedMap.remove(position)
        } else {
            selectedMap[position] = true
        }

        notifySelectedChange()
    }

    private fun notifySelectedChange() {
        onSelectedChangeListener?.onChange(
            selectedMap.filter { it.value }
                .map { getItem(it.key) }
        )
    }

    inner class RecyclerViewHolders(
        itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        val textView: TextView by lazy { itemView.findViewById<View>(R.id.textView) as TextView }
        val viewHeader: View by lazy { itemView.findViewById<View>(R.id.viewHeader) as View }
        val ivState by lazy { itemView.findViewById<ImageView>(R.id.ivState) }
        val cbDownload by lazy { itemView.findViewById<CheckBox>(R.id.cbDownload) }
        val progressBar by lazy { itemView.findViewById<ProgressBar>(R.id.progressBar) }

        fun setRead(read: Boolean) {
            val bannerColor = if (read) {
                ContextCompat.getColor(context, R.color.color_read_banner)
            } else {
                ContextCompat.getColor(context, R.color.color_unread_banner)
            }
            viewHeader.setBackgroundColor(bannerColor)
        }

        fun setDownloadState(state: DownloadState) {
            when (state) {
                NONE -> {
                    ivState.visibility = INVISIBLE
                    progressBar.visibility = INVISIBLE
                }
                FINISHED -> {
                    ivState.visibility = VISIBLE
                    progressBar.visibility = INVISIBLE
                    ivState.setImageResource(R.drawable.ic_checked)
                }
                PENDING -> {
                    ivState.visibility = VISIBLE
                    progressBar.visibility = INVISIBLE
                    ivState.setImageResource(R.drawable.ic_pending)
                }
                DOWNLOADING -> {
                    ivState.visibility = INVISIBLE
                    progressBar.visibility = VISIBLE
                    ivState.setImageResource(R.drawable.ic_downloading)
                }
                ERROR -> {
                    ivState.visibility = VISIBLE
                    progressBar.visibility = INVISIBLE
                    ivState.setImageResource(R.drawable.ic_error)
                }
            }
        }

        fun setSelectable(selectable: Boolean, selected: Boolean) {
            if (selectable) {
                cbDownload.visibility = VISIBLE
            } else {
                cbDownload.visibility = GONE
            }
            cbDownload.isChecked = selected
        }
    }

    interface OnItemClickListener {
        fun onClick(chapter: MangaChapterItem)
    }

    interface OnSelectedChangeListener {
        fun onChange(chapterList: List<MangaChapterItem>)
    }

    companion object {
        val diffCallBack = object : DiffUtil.ItemCallback<MangaChapterItem>() {
            override fun areItemsTheSame(oldItem: MangaChapterItem,
                newItem: MangaChapterItem): Boolean {
                return oldItem.hash == newItem.hash

            }

            override fun areContentsTheSame(oldItem: MangaChapterItem,
                newItem: MangaChapterItem): Boolean {
                return oldItem == newItem
            }
        }
    }
}
